### ISSUE TITLE
Process.CloseMainWindow does not work on MacOS

### ISSUE THREAD

**Issue by valentiniliescu** [dotnet/runtime#70014]
### Description

When executing `Process.CloseMainWindow` for an existing process on MacOS, it always returns false and the main window does not close.

Note that when executing the same code on Windows or on MacOS under Mono, `Process.CloseMainWindow` returns true and closes the window, as expected.

### Reproduction Steps

```
using System.Diagnostics;

// manually launch TextEdit before executing this code
foreach (var process in Process.GetProcessesByName("TextEdit"))
{
    Console.WriteLine($"Found process {process.ProcessName} with id {process.Id}");
    var closeResult = process.CloseMainWindow();
    Console.WriteLine($"Was main window closed: {closeResult}");
}
```

### Expected behavior

The code snippet prints out:
```
Found process TextEdit with id 46109
Was main window closed: True
```
and the TextEdit window is closed

### Actual behavior

The code snippet prints out:
```
Found process TextEdit with id 46109
Was main window closed: False
```
and the TextEdit window is still visible

### Regression?

It works fine when running under Mono 6.12.0.114 on the same MacOS version listed in the configuration.

### Configuration

.NET SDK (reflecting any global.json):
 Version:   6.0.202
 Commit: f8a55617d2

Runtime Environment:
 OS Name:     Mac OS X
 OS Version:  12.4
 OS Platform: Darwin
 RID:         osx.12-x64
 Base Path:   /usr/local/share/dotnet/sdk/6.0.202/

Host (useful for support):
  Version: 6.0.4
  Commit: be98e88c76

**Comment by valentiniliescu** [comment]
> Currently `CloseMainWindow` has no support for non-Windows platforms. Support may be ported from Mono.

In this case I would expect to get a `PlatformNotSupportedException`

**Comment by filipnavara** [comment]
> Support may be ported from Mono.

Mono doesn't quite implement it either. It basically does what `.Kill()` method does which exists in .NET 6 and is implemented on macOS.

**Comment by adamsitnik** [comment]
@stephentoub there are multiple `Process` APIs that are implemented only on Windows, they are not marked with `[SupportedOSPlatform("windows")]` and they do nothing or return defaul value for non-Windows configs:
* `CloseMainWindow`
* `EnterDebugMode`
* `LeaveDebugMode`
* `MainWindowHandle`
* `MainWindowTitle`

These methods don't seem to be hints like `ProcessThread.IdealProcessor` and `ProcessThread.I.ResetIdealProcessor`

Is there any reason for why we should not annotate them properly and make them throw PNSE?

**Comment by stephentoub** [comment]
EnterDebugMode and LeaveDebugMode have little use on anything other than Windows; you'd use them if you wanted to query for certain things that you can only do when you have the right privileges, and you can't query for those on other OSes, so these methods don't need to do anything special on those OSes.  Thus nop is a valid implementation.  We can mark them as unsupported if you want, but I expect that would only increase friction without actually helping anything.

MainWindowTitle returns string.Empty on Windows if it can't get a title.  So that's what we do on other OSes as well.  If we can enhance the implementations to do better on Linux/macOS, great, but changing it to throw PNSE and be unsupported seems inconsistent and doesn't help with much.

Same for MainWindowHandle... on Windows if it can't find one, it returns IntPtr.Zero.  Linux and macOS do the same.

For CloseMainWindow, it returns a bool indicating whether a main window could be successfully closed.  It can return false on Windows.  I'm also not seeing a big benefit of changing the behavior of the method to have an exceptional mode.

**Comment by adamsitnik** [comment]
@stephentoub how about just annotating these methods with `[SupportedOSPlatform("windows")]`? I just would like to make sure that our customers know about the fact that these methods are Windows-specific before they deploy their apps to other OSes and find out about it the hard way.

**Comment by stephentoub** [comment]
We can if you really think the benefits outweigh the cons.  It means code like
https://github.com/PowerShell/PowerShell/blob/ea3ea96cc75f51eab30e0bd93c7b4f8c31bbf14b/src/System.Management.Automation/engine/remoting/commands/EnterPSHostProcessCommand.cs#L815-L828
will start issuing warnings.  Does that bring enough benefit to justify it?  Up to you.

**Comment by mightypanda** [comment]
So, what is the suggested method to gracefully close a process with a UI window on MacOS ?

**Comment by huoyaoyuan** [comment]
> So, what is the suggested method to gracefully close a process with a UI window on MacOS ?

You can search for the native way to send the signal, and define P/Invoke.

**Comment by adamsitnik** [comment]
> You can search for the native way to send the signal, and define P/Invoke.

@huoyaoyuan is right, we don't have a dedicated managed API for sending signals to processes. 

BTW @stephentoub have we considered adding such API in the past? Perhaps when you were working on  https://github.com/dotnet/runtime/pull/55333 ?

For now, you can call the sys-call directly. Here is how we do it in System.CommandLine:

https://github.com/dotnet/command-line-api/blob/2c8df7382615b7b0d88e78a2a6205af0a8130eb0/src/System.CommandLine.Tests/Invocation/CancelOnProcessTerminationTests.cs#L103

https://github.com/dotnet/command-line-api/blob/2c8df7382615b7b0d88e78a2a6205af0a8130eb0/src/System.CommandLine.Tests/Invocation/CancelOnProcessTerminationTests.cs#L114-L115

**Comment by danmoseley** [comment]
This may be an alternative.

https://www.nuget.org/packages/Mono.Posix

**Comment by stephentoub** [comment]
> BTW @stephentoub have we considered adding such API in the past? Perhaps when you were working on https://github.com/dotnet/runtime/pull/55333 ?

Yes, it's been on our list to consider.  I think @tmds was thinking about what APIs would look like at one point.  And in the meantime there's the Mono.Posix nuget that Dan mentions.
